db/migrations/20190220152347_create_urns_query.sql

```diff
@@ -51,40 +51,45 @@ WITH urns AS (SELECT urns.id AS urn_id, ilks.id AS ilk_id, ilks.ilk, urns.identi
      inks AS ( -- Latest ink for each urn
          SELECT DISTINCT ON (urn_id) urn_id, ink, block_number
          FROM maker.vat_urn_ink
+                  LEFT JOIN public.headers ON vat_urn_ink.header_id = headers.id
          WHERE block_number <= all_urns.block_height
          ORDER BY urn_id, block_number DESC),
      arts AS ( -- Latest art for each urn
          SELECT DISTINCT ON (urn_id) urn_id, art, block_number
          FROM maker.vat_urn_art
+                  LEFT JOIN public.headers ON vat_urn_art.header_id = headers.id
          WHERE block_number <= all_urns.block_height
          ORDER BY urn_id, block_number DESC),
      rates AS ( -- Latest rate for each ilk
          SELECT DISTINCT ON (ilk_id) ilk_id, rate, block_number
          FROM maker.vat_ilk_rate
+                  LEFT JOIN public.headers ON vat_ilk_rate.header_id = headers.id
          WHERE block_number <= all_urns.block_height
          ORDER BY ilk_id, block_number DESC),
      spots AS ( -- Get latest price update for ilk. Problematic from update frequency, slow query?
          SELECT DISTINCT ON (ilk_id) ilk_id, spot, block_number
          FROM maker.vat_ilk_spot
+                  LEFT JOIN public.headers ON vat_ilk_spot.header_id = headers.id
          WHERE block_number <= all_urns.block_height
          ORDER BY ilk_id, block_number DESC),
      created AS (SELECT urn_id, api.epoch_to_datetime(block_timestamp) AS datetime
-                 FROM (SELECT DISTINCT ON (urn_id) urn_id, block_hash
+                 FROM (SELECT DISTINCT ON (urn_id) urn_id, block_timestamp
                        FROM maker.vat_urn_ink
-                       ORDER BY urn_id, block_number ASC) earliest_blocks
-                          LEFT JOIN public.headers ON hash = block_hash),
+                                LEFT JOIN public.headers ON vat_urn_ink.header_id = headers.id
+                       ORDER BY urn_id, block_number ASC) earliest_blocks),
      updated AS (SELECT DISTINCT ON (urn_id) urn_id, api.epoch_to_datetime(block_timestamp) AS datetime
-                 FROM ((SELECT DISTINCT ON (urn_id) urn_id, block_hash
```

Comment:
Good call! I think we can depend on `block_timestamp` being unique, but I also wonder if it might make sense to just select distinct on `header_id` 🤔

Reply:
Oh wait, think I misread - isn't this only selecting distinct on `urn_id`? And then it's just selecting the `block_timestamp` for each distinct `urn_id` so that it can use it in `api.epoch_to_datetime`